pyci/rdm/conditions.py

```diff
@@ -0,0 +1,66 @@
+# This file is part of PyCI.
+#
+# PyCI is free software: you can redistribute it and/or modify it under
+# the terms of the GNU General Public License as published by the Free
+# Software Foundation, either version 3 of the License, or (at your
+# option) any later version.
+#
+# PyCI is distributed in the hope that it will be useful, but WITHOUT
+# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+# FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License
+# for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with PyCI. If not, see <http://www.gnu.org/licenses/>.
+
+r"""PyCI RDM Conditions module."""
+
+import numpy as np
+
+__all__ = [
+    "find_closest_sdp",
+]
+
+def find_closest_sdp(dm, constraint, alpha):
+    r"""
+    Projection onto a semidefinite constraint.
+
+    Parameters
+    ----------
+    dm : np.ndarray
+        Density matrix.
+    constraint : function
+        Positive semidefinite constraint, linear mapping.
+    alpha : float
+        Value of the correct trace.
+
+    """
+    #symmetrize if necessary
+    L = constraint(dm) + constraint(dm).conj().T
+    #find eigendecomposition
+    vals, vecs = np.linalg.eig(L)
+    #calculate the shift, sigma0
+    sigma0 = calculate_shift(vals, alpha)
+    
+    #calculate the closest semidefinite positive matrix with correct trace
+    L_closest = vals @ np.diag(vecs - sigma0) @ vecs.conj().T
+
+    # return the reconstructed density matrix
+    return constraint(L_closest).conj().T
+
+
+def calculate_shift(eigenvalues, alpha):
+    r"""
+    Calculate the shift for density matrix for it to have the correct trace. 
+    This step shifts the spectrum and eliminates the negative eigenvalues.
+
+    Parameters
+    ----------
+    eigenvalues : np.ndarray
+        Eigenvalues of not shifted matrix.
+    alpha : float
+        Value of the coprrect trace.
+        
+    """
+    pass
```

Comment:
We can implement this by finding a root of a function $\alpha - \sum_{i}\Theta(\lambda _i - \sigma_0 )(\lambda_i - \sigma_0)$
It looks like a regular convex function to me so it should be any problems with it

the draft code would look can like this:
```
from scipy.optimize import root

trace = lambda sigma0: np.sum(np.heaviside(eigs - sigma0, 0.5)*(eigs-sigma0))
constraint = lambda x: trace(x) - alpha
res = root(constraint, 0)
```
I don't know if we should define the heaviside function as 0, 0.5, or 1. I prefer 0.5, but I don't know if this matters in this case
@PaulWAyers do you agree on this?

Reply:
The value of the Heaviside at zero should not matter; it gets multiplied by the eigenvalue zero anyway.

I think the optimal value of the shift has to be determined iteratively; that's the algorithm we used before. As it is a discontinuous funciton doing anything else is likely problematic. In any event, it is a super fast procedure; make it *understandable*. 

The algorithm we used before was very straightforward.
1. Set negative eigenvalues to zero.
2. Choose $\sigma_0$ to restore normalization.
3. Go to step 1. Continue until no eigenvalues are negative.